### PR TITLE
New way how to work with IN clause in OracleDB

### PR DESCRIPTION
**Pull request by stavamichal**
- we can use database type "array" - full supported just for oracle
 - method to work with sql array was added to DatabaseManagerBl|BlImpl
 - all getRequiredAttributesForBulk methods from AttributesManagerBlImpl
 were removed and logic was added directly to the methods in impl layer
 - all getRequiredAttributes methods in AttributesManagerImpl layer were
 redesigned to work with new oracle array and if they are unable to work
 with this specific desing, they will fallback on bulk of sql queries
 with maximum of objects in SQL IN CLAUSE defined by constant in code
 - use java reflection api to prevent oracle driver be part of perun

### CODE REVIEW

**Comment by stavamichal** [comment]
Changed as suggested @Vojtech-Sassmann 
Thank you.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 1]
```diff
@@ -3415,16 +3418,54 @@ public void deleteAllAttributeAuthz(PerunSession sess, AttributeDefinition attri
...
+			//This is possible if oracle sql array can't be created successfully (for example if oracle driver is missing)
+			//If happen, log it as error and continue using the standard way to obtain objects by IN clause
+			log.error("Unable to use Oracle SQL Array to obtain objects from database! Standard way to obtain these objects will be used instead.", ex);
+		}
+
+		HashMap<Member, List<Attribute>> allReturnedAttrs = new HashMap<>();
+		int from = 0;
+		int to = MAX_SIZE_OF_BULK_IN_SQL;
+
+		//Do not call one query for more than MAX_SIZE_OF_BULK_IN_SQL values in "SQL IN CLAUSE"
```
Comment (martin-kuba) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: The same trick as for Oracle can be done also for PostgreSQL, just use 
`WHERE id = ANY(?)` instead of `WHERE id IN (SELECT * FROM TABLE(?))` and use 
```
Array userIds = con.createArrayOf(JDBCType.INTEGER.name(), new Integer[] { 1,2,3 });
```
to create the array.

Comment (martin-kuba) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: The same can be done for HSQL, just use 
`WHERE id IN ( UNNEST(?) )` instead of `WHERE id = ANY(?)`

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 2]
```diff
@@ -3440,19 +3481,57 @@ public void deleteAllAttributeAuthz(PerunSession sess, AttributeDefinition attri
...
+		//Reason: SQL error in DB for too much records in one SQL query
+		do {
```
Comment (martin-kuba) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: again, for PostgreSQL and HSQL, the same trick with array can be used instead of generating IN clause with thousands of constants

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 3]
```diff
@@ -3469,17 +3548,56 @@ public void deleteAllAttributeAuthz(PerunSession sess, AttributeDefinition attri
...
+			log.error("Unable to use Oracle SQL Array to obtain objects from database! Standard way to obtain these objects will be used instead.", ex);
 		}
+
+		HashMap<User, List<Attribute>> allReturnedAttrs = new HashMap<>();
+		int from = 0;
+		int to = MAX_SIZE_OF_BULK_IN_SQL;
+
+		//Do not call one query for more than MAX_SIZE_OF_BULK_IN_SQL values in "SQL IN CLAUSE"
+		//Reason: SQL error in DB for too much records in one SQL query
+		do {
```
Comment (martin-kuba) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: again, PostgreSQL and HSQL with array

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 4]
```diff
@@ -3495,19 +3613,58 @@ public void deleteAllAttributeAuthz(PerunSession sess, AttributeDefinition attri
...
```
Comment (martin-kuba) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: again, PostgreSQL and HSQL with array

**Comment by stavamichal** [comment]
@martin-kuba It should be reworked and fixed now. Please check it again.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/DatabaseManagerBlImpl.java [hunk 5]
```diff
@@ -113,7 +123,40 @@ protected void initialize() throws InternalErrorException {
...
+			int[] arrayOfBeansIds = perunBeans.stream().mapToInt(PerunBean::getId).toArray();
+			try {
+				Method createOracleArrayMethod = connection.getClass().getMethod(nameOfOracleArrayMethod, String.class, Object.class);
+				createOracleArrayMethod.setAccessible(true);
+				return (java.sql.Array) createOracleArrayMethod.invoke(connection, AttributesManager.ORACLE_ARRAY_OF_NUMBERS, arrayOfBeansIds);
+			} catch (Exception ex) {
+				throw new InternalErrorRuntimeException("Can't access to method " + nameOfOracleArrayMethod, ex);
+			}
+		} else {
+			Integer[] arrayOfBeansIds = perunBeans.stream().map(PerunBean::getId).collect(Collectors.toList()).stream().toArray(Integer[]::new);
```
Comment (martin-kuba) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/DatabaseManagerBlImpl.java]: I believe it can be shorter:
```suggestion
			Integer[] arrayOfBeansIds = perunBeans.stream().map(PerunBean::getId).toArray(Integer[]::new);
```

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 6]
```diff
@@ -3469,15 +3487,23 @@ public void deleteAllAttributeAuthz(PerunSession sess, AttributeDefinition attri
...
+			return jdbc.execute("SELECT " + getAttributeMappingSelectQuery("usr_fac") + ", users.id FROM attr_names " +
+				"JOIN service_required_attrs ON attr_names.id=service_required_attrs.attr_id AND service_required_attrs.service_id=? " +
+				"JOIN users ON users.id " + Compatibility.getStructureForInClause() +
+				"LEFT JOIN user_facility_attr_values usr_fac ON attr_names.id=usr_fac.attr_id AND facility_id=? AND user_id=users.id " +
+				"WHERE namespace IN (?,?,?)", (PreparedStatementCallback<HashMap<User, List<Attribute>>>) preparedStatement -> {
+					Array sqlArray = ((PerunBlImpl)perun).getDatabaseManagerBl().prepareSQLArrayOfNumbers(users, preparedStatement);
+					preparedStatement.setInt(1, service.getId());
+					preparedStatement.setArray(2, sqlArray);
+					preparedStatement.setInt(3, facility.getId());
+					preparedStatement.setNString(4, AttributesManager.NS_USER_FACILITY_ATTR_DEF);
```
Comment (martin-kuba) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: calling setNString() will fail on PostgreSQL with SQLFeatureNotSupportedException, use setString() for all databases, the constants contain only ASCII characters

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java [hunk 7]
```diff
@@ -3495,17 +3521,24 @@ public void deleteAllAttributeAuthz(PerunSession sess, AttributeDefinition attri
...
-		} catch (RuntimeException ex) {
+			return jdbc.execute("SELECT " + getAttributeMappingSelectQuery("usr") + ", users.id FROM attr_names " +
+				"JOIN service_required_attrs on attr_names.id=service_required_attrs.attr_id AND service_required_attrs.service_id=? " +
+				"JOIN users ON users.id " + Compatibility.getStructureForInClause() +
+				"LEFT JOIN user_attr_values usr ON attr_names.id=usr.attr_id AND user_id=users.id " +
+				"WHERE namespace IN (?,?,?,?)", (PreparedStatementCallback<HashMap<User, List<Attribute>>>) preparedStatement -> {
+				Array sqlArray = ((PerunBlImpl) perun).getDatabaseManagerBl().prepareSQLArrayOfNumbers(users, preparedStatement);
+				preparedStatement.setInt(1, service.getId());
+				preparedStatement.setArray(2, sqlArray);
+				preparedStatement.setNString(3, AttributesManager.NS_USER_ATTR_CORE);
```
Comment (martin-kuba) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/AttributesManagerImpl.java]: dtto, use setString()

File: perun-core/src/main/resources/oracleChangelog.txt [hunk 8]
```diff
@@ -6,6 +6,11 @@
 -- Directly under version number should be version commands. They will be executed in the order they are written here.
 -- Comments are prefixed with -- and can be written only between version blocks, that means not in the lines with commands. They have to be at the start of the line.
 
+3.1.52
+create or replace type TARRAYOFCHARACTERS is table of  nvarchar2(100);
```
Comment (zlamalp) [inline perun-core/src/main/resources/oracleChangelog.txt]: Maybe we don't need to restrict ourselves with limits to character length or number size ?, also unnecessary whitespace.

File: perun-db/oracle.sql [hunk 9]
```diff
@@ -1805,7 +1805,7 @@ CREATE INDEX ufauv_idx ON user_facility_attr_u_values (user_id, facility_id, att
 CREATE INDEX vauv_idx ON vo_attr_u_values (vo_id, attr_id) ;
 
 -- set initial Perun DB version
-insert into configurations values ('DATABASE VERSION','3.1.51');
+insert into configurations values ('DATABASE VERSION','3.1.52');
```
Comment (zlamalp) [inline perun-db/oracle.sql]: Please update also version in a comment at the top of the file.

File: perun-db/postgres.sql [hunk 10]
```diff
@@ -1903,7 +1903,7 @@ grant all on user_ext_source_attr_u_values to perun;
 grant all on members_sponsored to perun;
 
 -- set initial Perun DB version
-insert into configurations values ('DATABASE VERSION','3.1.51');
+insert into configurations values ('DATABASE VERSION','3.1.52');
```
Comment (zlamalp) [inline perun-db/postgres.sql]: Please update also version in a comment at the top of the file.

**Comment by stavamichal** [comment]
I would continue with other login in next PR if it's alright with you. What do you think? Other thing should be fixed.
